Source/DafnyLanguageServer/Workspace/Notifications/VerificationDiagnosticsParams.cs

```diff
@@ -0,0 +1,590 @@
+        } else {
+          previousNotUnknown = result[line];
+        }
+      }
+
+      var resolutionErrorRendered = 0;
+      foreach (var diagnostic in diagnostics) {
+        if (resolutionErrorRendered >= numberOfResolutionErrors) {
+          break;
+        }
+        if (diagnostic.Range.Start.Line >= 0 && diagnostic.Range.Start.Line < result.Length) {
+          result[diagnostic.Range.Start.Line] = LineVerificationStatus.ResolutionError;
+          resolutionErrorRendered++;
+        }
+      }
+
+      var existsErrorRange = false;
+      var existsError = false;
+      foreach (var line in result) {
+        existsErrorRange = existsErrorRange || line == LineVerificationStatus.ErrorContext;
+        existsError = existsError || line == LineVerificationStatus.AssertionFailed;
+      }
+
+      if (existsErrorRange && !existsError) {
+        existsError = false;
+      }
+
+      return result;
+    }
+  }
+
+
+  public enum VerificationStatus {
+    Nothing = 0,
+    Verified = 200,
+    Inconclusive = 270,
+    Error = 400
+  }
+
+  public enum CurrentStatus {
+    Current = 0,
+    Obsolete = 1,
+    Verifying = 2
+  }
+
+  public enum LineVerificationStatus {
+    // Default value for every line, before the renderer figures it out.
+    Nothing = 0,
+    // For first-time computation not actively computing but soon. Synonym of "obsolete"
+    // (scheduledComputation)
+    Scheduled = 1,
+    // For first-time computations, actively computing
+    Verifying = 2,
+    VerifiedObsolete = 201,
+    VerifiedVerifying = 202,
+    // Also applicable for empty spaces if they are not surrounded by errors.
+    Verified = 200,
+    // For trees containing children with errors (e.g. functions, methods, fields, subset types)
+    ErrorContextObsolete = 301,
+    ErrorContextVerifying = 302,
+    ErrorContext = 300,
+    // For individual assertions in error contexts
+    AssertionVerifiedInErrorContextObsolete = 351,
+    AssertionVerifiedInErrorContextVerifying = 352,
+    AssertionVerifiedInErrorContext = 350,
+    // For specific lines which have errors on it. They take over verified assertions
+    AssertionFailedObsolete = 401,
+    AssertionFailedVerifying = 402,
+    AssertionFailed = 400,
+    // For lines containing resolution or parse errors
+    ResolutionError = 500
+  }
+
+  /// <summary>
+  /// A verification tree is an abstraction over the code to represent the verification
+  /// status of a region of the document, useful for IDE verification inspection.
+  /// A verification tree can contain other child trees.
+  /// It can currently be rendered linearly, e.g. for gutter display, or used as a tree in a test-like display. 
+  /// The verification status consists of two orthogonal concepts:
+  /// - StatusVerification: Nothing (initial), Error, Verified, or Inconclusive
+  /// - StatusCurrent: Current (Up-to-date), Obsolete (outdated), and Verifying (as notified by the verifier)
```

Comment:
I don't think these are orthogonal, especially if you agree to drop Obsolete. You just end up with `Nothing (initial), Verifying, Error, Verified, or Inconclusive`

Reply:
> I don't think there's value in knowing that a line WAS correct/invalid previously if it may or may not change once verification catches up (either automatically or once you save).

The problem of removing the previous verification result is that any given line will transition is that, during the time it verifies (which are a few seconds), what you suggests really means hiding the previous state during the time Dafny verifies. Why would you hide the previous state way before you have some more up-to-date information? Users trying to fix an error might be thinking in parallel about the code and the previous verification result in other places which they believe would not have been impacted. Removing them this information would result in a frustrating experience.

There are precedents about that: We decided to still display previous verification diagnostics (migrate them) although there might be parse/resolution errors. Moreover, in general, diagnostics are "updated" rather than "removed" and then "re-added" a few seconds later.
So, in short, it's something that users would expect and it's desirable.

To illustrate my point, let's consider a CALC statemen with four statements, where Q2 is failing but Q3 and Q4 are succeeding. A user makes an edit on Q2, which makes it verify, but now Q3 does not verify anymore. The trace under this PR would be the following:
```dafny
[ ][I][S][ ]: calc {
[ ][I][S][ ]:   Q1;
[=][-][~][O]:   Q2;
[O][o][Q][=]:   Q3;
[O][o][Q][O]:   Q4;
[ ][I][S][ ]: }
```

With your approach of not making statuses orthogonal, I don't think how it would look better than this:
```dafny
[ ] I  S [ ]: calc {
[ ] I  S [ ]:   Q1;
[=] I  S [O]:   Q2;
[O] I  S [=]:   Q3;
[O] I  S [O]:   Q4;
[ ] I  S [ ]: }
```

Here are notable differences in the two approaches:
- In the first version, the switching state from "error" to "verified" is instant on the fourth update (letting the user know the error was first obsolete, and then recomputing), whereas in the second version, there is an intermediate state that completely hides all the previous information.
- If I did not pay attention about the verification result between Q3 and Q4 (which can be farther away) because I was focused on Q1 to Q2, during the whole time the verifier runs (a few seconds), I would be blind about having reasons to believe that Q3 is still equals to Q4.
- A few seconds of completely removing the previous information would have the effect to remove alignment information for the eye, so instead of seeing if an error has been fixed and the user can move on, the eye will have to start its course again from the line being edited to the left where the gutter is.
- During the UX development with a UX designer, she agreed with me that it's better to minimize the frequency of full visual changes made to the gutter, because it should be supportive for the eye, not distract it.
- Ideally, the eye should be able to stay on the code and only use peripheral vision to know the verification status. With this PR, the eye will be able to see an error "going down" with the peripheral vision, and since the user previously saw that this error was on the line being edited, they won't even need to look from right to left to see if the line has an error or not. With your alternative, after few seconds during which that information disappears, the eye will not be able to use peripheral vision to determine if a line verifies or not until it does the actual movement from right to left.

I can go on and on, and also explain you why displaying the error context is important as well, rather than just repeating the verification status on every line. Just ask me.